### PR TITLE
refactor: switch to interpreter

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR is the final one that switches the old simplifier function out with the interpreter.

**Related issue(s)**
fixes https://github.com/asyncapi/generator-model-sdk/issues/64

### CODE REVIEW

**Comment by jonaslagoni** [comment]
@czlowiek488 would love to get your feedback on the documentation of the interpreter if you have time 🙂 Can't request a review from you for some reason, so this will have to do 😄

File: docs/interpretation_of_JSON_Schema_draft_7.md
Comment (magicmatatjahu) [inline docs/interpretation_of_JSON_Schema_draft_7.md:51]:
```suggestion
Because of the recursive nature of the interpreter (and the nested nature of JSON Schema) it happens that two models needs to be merged together.
```

File: docs/interpretation_of_JSON_Schema_draft_7.md
Comment (magicmatatjahu) [inline docs/interpretation_of_JSON_Schema_draft_7.md:54]:
```suggestion
- `additionalProperties` if both models contain items they are recursively merged together.
```

File: docs/interpretation_of_JSON_Schema_draft_7.md
Comment (magicmatatjahu) [inline docs/interpretation_of_JSON_Schema_draft_7.md:7]:
```suggestion
We only provide the underlying structure of the schema file for the model, where constraints/annotations such as `maxItems`, `uniqueItems`, `multipleOf`, etc, are not transformed.
```

File: docs/interpretation_of_JSON_Schema_draft_7.md
Comment (magicmatatjahu) [inline docs/interpretation_of_JSON_Schema_draft_7.md:62]: We have information about `originalSchema` only in `API.md` file, we should described what is `originalSchema` here in another place :)

**Review by magicmatatjahu — COMMENTED** [review]
Some comments :)

File: docs/interpretation_of_JSON_Schema_draft_7.md
Comment (jonaslagoni) [reply docs/interpretation_of_JSON_Schema_draft_7.md:54]: Where does `items` fit in? 🤔

Comment (magicmatatjahu) [reply docs/interpretation_of_JSON_Schema_draft_7.md:54]: Sorry 😅 My fault. I did a review right after waking up. It can stay as is, it makes more sense :)

File: docs/interpretation_of_JSON_Schema_draft_7.md
Comment (jonaslagoni) [reply docs/interpretation_of_JSON_Schema_draft_7.md]: Hmm, yea true. What do you think is the best solution? 🤔

Comment (magicmatatjahu) [reply docs/interpretation_of_JSON_Schema_draft_7.md]: It can be described here, or somewhere at the beginning of this doc :)

Comment (jonaslagoni) [reply docs/interpretation_of_JSON_Schema_draft_7.md]: Added a link to the input processor docs which explains the CommonModel, otherwise, we have to keep in mind that we have a description of a property in this document as well.

**Review by magicmatatjahu — APPROVED** [review]

File: src/processors/JsonSchemaInputProcessor.ts
Comment (czlowiek488) [inline src/processors/JsonSchemaInputProcessor.ts:215]: I guess interpreter should be constructed before this method execution. Please remember that creating instance by using keyword `new` is not well optimized in javascript and should be avoided at execution time.

File: test/processors/JsonSchemaInputProcessor.spec.ts
Comment (czlowiek488) [inline test/processors/JsonSchemaInputProcessor.spec.ts:10]: I should blame you every single time you use `let` or mutate something in your code ;)

File: src/interpreter/Interpreter.ts
Comment (czlowiek488) [inline src/interpreter/Interpreter.ts:110]: For me you violate YAGNI, KISS and SRP. And I will point it up whenever I see someting like that. If you would be my collegue at work I would block your PR.

**Review by czlowiek488 — CHANGES_REQUESTED** [review]
Looks fine but I would change few things. You may be fine with this codebase. However I am not 😞

File: src/interpreter/Interpreter.ts
Comment (jonaslagoni) [reply src/interpreter/Interpreter.ts:110]: Hehe, I mean you kinda are, let me deep dive a bit and try and break it up to better understand it

**Comment by jonaslagoni** [comment]
> Looks fine but I would change few things. You may be fine with this codebase. However I am not 😞

Thanks for the review @czlowiek488, gonna go through your comments now 💪

File: test/processors/JsonSchemaInputProcessor.spec.ts
Comment (jonaslagoni) [reply test/processors/JsonSchemaInputProcessor.spec.ts:10]: What do you mean 😄 ? 

This is the only way I found that could alter the mocked return, damn Jest..

File: src/processors/JsonSchemaInputProcessor.ts
Comment (jonaslagoni) [reply src/processors/JsonSchemaInputProcessor.ts:215]: We could just have one interpreter instance where we could reset the iterated models, but it's an optimization we can make later if it ever becomes a problem.

File: src/interpreter/Interpreter.ts
Comment (jonaslagoni) [reply src/interpreter/Interpreter.ts:110]: I agree with you here, it should be split in this case. Changed 👍

**Comment by jonaslagoni** [comment]
So caught 2 problems with the current interpreter.
1. The blackbox testing actually caught an error with how the separate models got saved, had to add an integration test to ensure that models are iterated and returned correctly. Unit testing with mocks would be too complex, so switched to that.
2. I forgot to implement split functionality for `patternProperties`, `additionalProperties`, and `items`, so I added tests and implementation.

File: src/interpreter/Interpreter.ts
Comment (magicmatatjahu) [inline src/interpreter/Interpreter.ts:141]:
```suggestion
  private trySplitModels(model: CommonModel): CommonModel {
```

File: src/interpreter/Interpreter.ts
Comment (magicmatatjahu) [inline src/interpreter/Interpreter.ts:157]:
```suggestion
  ensureModelsAreSplitted(model: CommonModel): void {
```

**Review by magicmatatjahu — COMMENTED** [review]
Only two changes :)

File: src/interpreter/Interpreter.ts
Comment (jonaslagoni) [reply src/interpreter/Interpreter.ts:157]: This sounds weird? "splitted" are not a word 🙂

File: src/interpreter/Interpreter.ts
Comment (jonaslagoni) [reply src/interpreter/Interpreter.ts:141]: I don't know why, but this sounds a bit weird as well 🤔? Does this make more sense for you?

File: src/interpreter/Interpreter.ts
Comment (magicmatatjahu) [reply src/interpreter/Interpreter.ts:157]: 😅 I was wondering too but I found the use of this word in some articles ... my mistake.

File: src/interpreter/Interpreter.ts
Comment (magicmatatjahu) [reply src/interpreter/Interpreter.ts:141]: My problem is `tryAnd...`, but `try` what? 😄

Comment (jonaslagoni) [reply src/interpreter/Interpreter.ts:141]: "try split models" vs "try and split models". When I read the name of the function the latter sounds more natural vs the first one 🤔

Comment (magicmatatjahu) [reply src/interpreter/Interpreter.ts:141]: As you want :) I don't think that we should use `And` and `Or` in function and variable names.

Comment (jonaslagoni) [reply src/interpreter/Interpreter.ts:141]: Yea true.

Comment (jonaslagoni) [reply src/interpreter/Interpreter.ts:141]: Sure lets change it 😄

Comment (jonaslagoni) [reply src/interpreter/Interpreter.ts]: Changed

Comment (magicmatatjahu) [reply src/interpreter/Interpreter.ts]: Ok, If you want, let's wait for last review from czlowiek488 :)

**Comment by jonaslagoni** [comment]
@czlowiek488 do you want to do a final review as well?

**Review by magicmatatjahu — APPROVED** [review]
LGTM! :)

File: test/processors/JsonSchemaInputProcessor.spec.ts
Comment (czlowiek488) [reply test/processors/JsonSchemaInputProcessor.spec.ts:10]: @jonaslagoni I mean that you should write inmutable code.

**Comment by jonaslagoni** [comment]
@czlowiek488  with no specific reply whether you wanted to review it, I am merging this PR now 🙂

**Review by czlowiek488 — COMMENTED** [review]
It is fine for me

**Pull request merged**
